sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala

```diff
@@ -251,6 +251,25 @@ case class ExpressionEncoder[T](
 
     val plan = Project(Alias(unbound, "")() :: Nil, LocalRelation(schema))
     val analyzedPlan = SimpleAnalyzer.execute(plan)
+    if (!analyzedPlan.resolved) {
+      analyzedPlan match {
+        // Looked for UpCasts. If these still exist, it indicates they could not resolve and
+        // something about the schemas is mismatched.
+        case p @ Project(projectList, _) => {
```

Comment:
Is there a reason to not just put this in `CheckAnalysis` and make it more general (i.e. not just `Project`)?  I realize its only used here now, but if we ever do use it elsewhere, its invalid to remain after analysis.